### PR TITLE
checkpoint refactor to save disk space

### LINKED ISSUE #674: save k-best ckpt
参考https://github.com/PaddlePaddle/Parakeet/pull/114 ，将checkpoint相关操作抽象成类，组合KBest实现功能。

### PR DESCRIPTION
**Pull request by 745165806**
### PR changes
修改checkpoint.py, 变为class, 并增加save k best, 和 save n last 属性

### Describe
fix #674

### CODE REVIEW

File: examples/tiny/s0/conf/deepspeech2.yaml [hunk 1]
```diff
@@ -43,12 +43,15 @@ model:
...
 
 training:
-  n_epoch: 24
+  n_epoch: 6
   lr: 1e-5 
   lr_decay: 1.0 
   weight_decay: 1e-06
   global_grad_clip: 5.0
   log_interval: 1
+  max_epoch: 3
```
Comment (zh794390558) [inline examples/tiny/s0/conf/deepspeech2.yaml]: 这两个参数表意不明。

Comment (745165806) [reply examples/tiny/s0/conf/deepspeech2.yaml]: 我改一下，加一个checkpoint小分支，然后max_size 和last_size

File: deepspeech/training/trainer.py [hunk 2]
```diff
@@ -18,7 +18,7 @@
 from paddle import distributed as dist
 from tensorboardX import SummaryWriter
 
-from deepspeech.utils import checkpoint
+from deepspeech.utils.checkpoint import KBestCheckpoint
```
Comment (zh794390558) [inline deepspeech/training/trainer.py]: KBestCheckpoint -> Checkpoint

File: deepspeech/utils/checkpoint.py [hunk 3]
```diff
@@ -23,130 +23,226 @@
...
 
+import glob
+
 logger = Log(__name__).getlog()
 
 __all__ = ["load_parameters", "save_parameters"]
 
+class KBestCheckpoint(object):
+    def __init__(self,
+                 max_size: int=5,
```
Comment (zh794390558) [inline deepspeech/utils/checkpoint.py]: max_size -> kbest_n, latest_n

File: deepspeech/utils/checkpoint.py [hunk 4]
```diff
@@ -23,130 +23,226 @@
 from deepspeech.utils import mp_tools
 from deepspeech.utils.log import Log
 
+import glob
+
 logger = Log(__name__).getlog()
 
 __all__ = ["load_parameters", "save_parameters"]
```
Comment (zh794390558) [inline deepspeech/utils/checkpoint.py]: Checkpoint

File: deepspeech/utils/checkpoint.py [hunk 5]
```diff
@@ -23,130 +23,226 @@
...
 
+class KBestCheckpoint(object):
+    def __init__(self,
+                 max_size: int=5,
+                 last_size: int=1):
+        self.best_records: Mapping[Path, float] = {}
```
Comment (zh794390558) [inline deepspeech/utils/checkpoint.py:35]: Mapping  Path 没有导入

File: deepspeech/utils/checkpoint.py [hunk 6]
```diff
@@ -23,130 +23,226 @@
...
+        self.max_size = max_size
+        self.last_size = last_size
+        self._save_all = (max_size == -1)
+    
+    def should_save_best(self, metric: float) -> bool:
+        if not self.best_full():
+            return True
+
+        # already full
+        worst_record_path = max(self.best_records, key=self.best_records.get)
```
Comment (zh794390558) [inline deepspeech/utils/checkpoint.py:110]: import operator
stats = {'a':1000, 'b':3000, 'c': 100}
max(stats.iteritems(), key=operator.itemgetter(1))[0]
使用上面的语法更清晰些？

File: deepspeech/utils/checkpoint.py [hunk 7]
```diff
@@ -23,130 +23,226 @@
...
+    def del_checkpoint(self, checkpoint_dir, tag_or_iteration):
+        checkpoint_path = os.path.join(checkpoint_dir,
+                                    "{}".format(tag_or_iteration))
+        for filename in glob.glob(checkpoint_path+".*"):
+            os.remove(filename)
+            print("delete file: "+filename)
+        
+
+
+    def _load_latest_checkpoint(self, checkpoint_dir: str) -> int:
```
Comment (zh794390558) [inline deepspeech/utils/checkpoint.py]: 抽取公共接口，实现load_latest, load_best

File: deepspeech/utils/checkpoint.py [hunk 8]
```diff
@@ -23,130 +23,226 @@
...
+                        checkpoint_dir, tag_or_iteration,
+                        model, optimizer, infos):
+        if("val_loss" not in infos.keys()):
+            self.save_parameters(checkpoint_dir, tag_or_iteration, 
+                                model, optimizer, infos)
+            return
+
+        #save best
+        if self.should_save_best(infos["val_loss"]):
+            self.save_checkpoint_and_update(infos["val_loss"],
```
Comment (zh794390558) [inline deepspeech/utils/checkpoint.py]: save_best_checkpoint_and_update

File: deepspeech/utils/checkpoint.py [hunk 9]
```diff
@@ -23,130 +23,226 @@
...
+
+    def best_full(self):
+        return (not self._save_all) and len(self.best_records) == self.max_size
+    
+    def last_full(self):
+        return len(self.last_records) == self.last_size
+
+    def add_checkpoint(self,
+                        checkpoint_dir, tag_or_iteration,
+                        model, optimizer, infos):
```
Comment (zh794390558) [inline deepspeech/utils/checkpoint.py]: 加个metric参数，default='val_loss"

File: deepspeech/utils/checkpoint.py [hunk 10]
```diff
@@ -23,130 +23,226 @@
...
+                        model, optimizer, infos)
+        self.best_records[tag_or_iteration] = metric
+ 
+    def save_last_checkpoint_and_update(self, checkpoint_dir, tag_or_iteration,
+                        model, optimizer, infos):
+        # remove the old
+        if self.last_full():
+            to_del_fn = self.last_records.pop(0)
+            if(to_del_fn not in self.best_records.keys()):
+                print('----to remove (last)----')
```
Comment (zh794390558) [inline deepspeech/utils/checkpoint.py]: print 替换成 logger.info

File: deepspeech/training/trainer.py [hunk 11]
```diff
@@ -139,9 +139,12 @@ def save(self, tag=None, infos: dict=None):
             "epoch": self.epoch,
             "lr": self.optimizer.get_lr()
         })
-        checkpoint.save_parameters(self.checkpoint_dir, self.iteration
+        self.checkpoint.add_checkpoint(self.checkpoint_dir, self.iteration
```
Comment (zh794390558) [inline deepspeech/training/trainer.py:142]: 测试多卡时是否只有rank在保存ckpt

File: deepspeech/utils/checkpoint.py (hunk 9 above)
Comment (745165806) [reply deepspeech/utils/checkpoint.py]: metric 在infos里面infos["val_loss"]

File: deepspeech/utils/checkpoint.py (hunk 6 above)
Comment (745165806) [reply deepspeech/utils/checkpoint.py:110]: 这个会报错，显示dic没有啥啥属性，所有没改

File: deepspeech/training/trainer.py [hunk 12]
```diff
@@ -64,7 +64,7 @@ class Trainer():
         The parsed command line arguments.
     Examples
     --------
-    >>> def main_sp(config, args):
+    >>> def p(config, args):
```
Comment (zh794390558) [inline deepspeech/training/trainer.py]: ？

Comment (745165806) [reply deepspeech/training/trainer.py]: 手误

File: deepspeech/utils/checkpoint.py [hunk 13]
```diff
@@ -141,21 +142,22 @@ def save_checkpoint_record(self, checkpoint_dir: str, iteration: int):
...
         with open(checkpoint_record_latest, "w") as handle:
             for i in self.latest_records:
                 handle.write("model_checkpoint_path:{}\n".format(i))
 
-
-    def load_last_parameters(self, model,
-                        optimizer=None,
-                        checkpoint_dir=None,
-                        checkpoint_path=None):
+    def load_last_parameters(self,
```
Comment (zh794390558) [inline deepspeech/utils/checkpoint.py]: load_latest_parameters

和load_best_parameters重复的多，可以抽取出一个公共的接口。

Comment (zh794390558) [reply deepspeech/utils/checkpoint.py]: 不对外的接口前加上下划线吧， 如 def __load_parameters

File: deepspeech/utils/checkpoint.py [hunk 14]
```diff
@@ -261,7 +210,7 @@ def load_best_parameters(self,
         return configs
 
     @mp_tools.rank_zero_only
-    def save_parameters(self,
```
Comment (zh794390558) [inline deepspeech/utils/checkpoint.py]: 对外部暴露的接口不用加下划线

File: deepspeech/utils/checkpoint.py [hunk 15]
```diff
@@ -22,131 +24,225 @@
 
 from deepspeech.utils import mp_tools
 from deepspeech.utils.log import Log
+# import operator
```
Comment (zh794390558) [inline deepspeech/utils/checkpoint.py]: delete

File: examples/tiny/s0/conf/deepspeech2.yaml [hunk 16]
```diff
@@ -43,12 +43,16 @@ model:
   share_rnn_weights: True 
 
 training:
-  n_epoch: 24
```
Comment (zh794390558) [inline examples/tiny/s0/conf/deepspeech2.yaml:46]: example下的所有conf都要对应修改。
aishell和librispeech中kbeat_n默认设置50， latest_n默认设置5吧

**Review by zh794390558 — APPROVED** [review]
LGTM
